airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java

```diff
@@ -549,6 +566,30 @@ private static void validateSchema(final RecordSchemaValidator recordSchemaValid
     }
   }
 
+  private static void populatedStreamToSelectedFields(ConfiguredAirbyteCatalog catalog,
+                                                      Map<AirbyteStreamNameNamespacePair, List<String>> streamToSelectedFields) {
+    for (final var s : catalog.getStreams()) {
+      final List<String> selectedFields = new ArrayList<>();
+      final JsonNode propertiesNode = s.getStream().getJsonSchema().findPath("properties");
+      if (propertiesNode.isObject()) {
+        propertiesNode.fieldNames().forEachRemaining((fieldName) -> selectedFields.add(fieldName));
+      } else {
+        throw new RuntimeException("No properties node in stream schema");
+      }
+      streamToSelectedFields.put(AirbyteStreamNameNamespacePair.fromConfiguredAirbyteSteam(s), selectedFields);
+    }
+  }
+
+  private static void filterSelectedFields(Map<AirbyteStreamNameNamespacePair, List<String>> streamToSelectedFields, AirbyteMessage airbyteMessage) {
+    final AirbyteRecordMessage record = airbyteMessage.getRecord();
+    final AirbyteStreamNameNamespacePair messageStream = AirbyteStreamNameNamespacePair.fromRecordMessage(record);
+    final List<String> selectedFields = streamToSelectedFields.getOrDefault(messageStream, Collections.emptyList());
+    final JsonNode data = record.getData();
+    if (data.isObject()) {
```

Comment:
This could be false in a case where record looks like `{"data": "a-string-and-not-an-object"}`. Any simple type would meet the requirement that it's a non-null JsonNode.

However, the docs say: `The data contains the record data and must always be populated. It is a JSON blob.` I think it's fair to interpret "JSON blob" as object here; so I'm comfortable throwing a RuntimeException if it happens.

Reply:
Ah hm, I forgot that a plain String could be a valid JsonNode. I wonder if that's a valid use case? Maybe we can ask someone who is more familiar with the Airbyte protocol if it seems like a good idea to throw a RuntimeException if we encounter a string in a record's `data` field. @gosusnp perhaps?

Reply:
The source acceptance tests seem to confirm that records must always have some key/value pairs: see https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py#L475 and https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/utils/json_schema_helper.py#L147 (though I could well be reading the tests wrong).